### ISSUE TITLE
[Question] Why are the training results of RL-Games and SKRL very different?

### ISSUE THREAD

**Issue by KevinGeng3** [isaac-sim/IsaacLab#1551]
### Question
I am currently working on single-arm reaching task. With almost the same parameters(well, I think they are almost the same) and the same envs, RL_Games easily achieves good training results while SKRL can not complete the task. Here are the trainning parameters.

RL_Games:
```
params:
  seed: 42

  # environment wrapper clipping
  env:
    # added to the wrapper
    clip_observations: 100.0
    # can make custom wrapper?
    clip_actions: 100.0

  algo:
    name: a2c_continuous

  model:
    name: continuous_a2c_logstd

  # doesn't have this fine grained control but made it close
  network:
    name: actor_critic
    separate: False
    space:
      continuous:
        mu_activation: None
        sigma_activation: None

        mu_init:
          name: default
        sigma_init:
          name: const_initializer
          val: 0
        fixed_sigma: True
    mlp:
      units: [64, 64]
      activation: elu
      d2rl: False

      initializer:
        name: default
      regularizer:
        name: None

  load_checkpoint: False # flag which sets whether to load the checkpoint
  load_path: '' # path to the checkpoint to load

  config:
    name: single_arm_reach_direct
    env_name: rlgpu
    device: 'cuda:0'
    device_name: 'cuda:0'
    multi_gpu: False
    ppo: True
    mixed_precision: False
    normalize_input: True
    normalize_value: True
    num_actors: -1  # configured from the script (based on num_envs)
    reward_shaper:
      scale_value: 0.1
    normalize_advantage: True
    gamma: 0.99
    tau : 0.95
    learning_rate: 5e-4
    lr_schedule: adaptive
    kl_threshold: 0.008
    score_to_win: 20000
    max_epochs: 1000
    save_best_after: 50
    save_frequency: 100
    grad_norm: 1.0
    entropy_coef: 0.01
    truncate_grads: True
    e_clip: 0.2
    horizon_length: 36
    minibatch_size: 73728
    mini_epochs: 5
    critic_coef: 2
    clip_value: True
    clip_actions: False
    bounds_loss_coef: 0.0001
```
SKRL:
```
seed: 42


# Models are instantiated using skrl's model instantiator utility
# https://skrl.readthedocs.io/en/latest/api/utils/model_instantiators.html
models:
  separate: False
  policy:  # see gaussian_model parameters
    class: GaussianMixin
    clip_actions: False
    clip_log_std: True
    min_log_std: -20.0
    max_log_std: 2.0
    initial_log_std: 0.0
    network:
      - name: net
        input: STATES
        layers: [64, 64]
        activations: elu
    output: ACTIONS
  value:  # see deterministic_model parameters
    class: DeterministicMixin
    clip_actions: False
    network:
      - name: net
        input: STATES
        layers: [64, 64]
        activations: elu
    output: ONE


# Rollout memory
# https://skrl.readthedocs.io/en/latest/api/memories/random.html
memory:
  class: RandomMemory
  memory_size: -1  # automatically determined (same as agent:rollouts)


# PPO agent configuration (field names are from PPO_DEFAULT_CONFIG)
# https://skrl.readthedocs.io/en/latest/api/agents/ppo.html
agent:
  class: PPO
  rollouts: 32
  learning_epochs: 8
  mini_batches: 2
  discount_factor: 0.99
  lambda: 0.95
  learning_rate: 3.0e-04
  learning_rate_scheduler: KLAdaptiveLR
  learning_rate_scheduler_kwargs:
    kl_threshold: 0.008
  state_preprocessor: RunningStandardScaler
  state_preprocessor_kwargs: null
  value_preprocessor: RunningStandardScaler
  value_preprocessor_kwargs: null
  random_timesteps: 0
  learning_starts: 0
  grad_norm_clip: 1.0
  ratio_clip: 0.2
  value_clip: 0.2
  clip_predicted_values: True
  entropy_loss_scale: 0.05
  value_loss_scale: 2.0
  kl_threshold: 0.008
  rewards_shaper_scale: 1.0
  time_limit_bootstrap: False
  # logging and checkpoint
  experiment:
    directory: "single_arm_reach_direct"
    experiment_name: ""
    write_interval: 100
    checkpoint_interval: 1000


# Sequential trainer
# https://skrl.readthedocs.io/en/latest/api/trainers/sequential.html
trainer:
  class: SequentialTrainer
  timesteps: 36000
  environment_info: log
```
I know that RL models are very sensitive to parameters, so could anyone help me find out some differences between them? Thank you!

**Comment by YandongJi** [comment]
It looks like setting time_limit_bootstrap to True can make training better for me

**Comment by Toni-SM** [comment]
Hi @KevinGeng3 

In the [Isaac Lab examples of the **skrl** docs](https://skrl.readthedocs.io/en/latest/intro/examples.html#isaac-lab-environments) you can find the appropriate mapping between *rl_games* and *skrl* configurations.

<hr>

This would be the mapping based on the *rl_games* configuration you present.
Please, adjust `mini_batches` according to the number of environments you are training on

```yaml
seed: 42


# Models are instantiated using skrl's model instantiator utility
# https://skrl.readthedocs.io/en/latest/api/utils/model_instantiators.html
models:
  separate: False
  policy:  # see gaussian_model parameters
    class: GaussianMixin
    clip_actions: False
    clip_log_std: True
    min_log_std: -20.0
    max_log_std: 2.0
    initial_log_std: 0.0
    network:
      - name: net
        input: STATES
        layers: [64, 64]
        activations: elu
    output: ACTIONS
  value:  # see deterministic_model parameters
    class: DeterministicMixin
    clip_actions: False
    network:
      - name: net
        input: STATES
        layers: [64, 64]
        activations: elu
    output: ONE


# Rollout memory
# https://skrl.readthedocs.io/en/latest/api/memories/random.html
memory:
  class: RandomMemory
  memory_size: -1  # automatically determined (same as agent:rollouts)


# PPO agent configuration (field names are from PPO_DEFAULT_CONFIG)
# https://skrl.readthedocs.io/en/latest/api/agents/ppo.html
agent:
  class: PPO
  rollouts: 36
  learning_epochs: 5
  mini_batches: 36 * NUM_ENVS / 73728  # horizon_length * num_actors / minibatch_size
  discount_factor: 0.99
  lambda: 0.95
  learning_rate: 5.0e-04
  learning_rate_scheduler: KLAdaptiveLR
  learning_rate_scheduler_kwargs:
    kl_threshold: 0.008
  state_preprocessor: RunningStandardScaler
  state_preprocessor_kwargs: null
  value_preprocessor: RunningStandardScaler
  value_preprocessor_kwargs: null
  random_timesteps: 0
  learning_starts: 0
  grad_norm_clip: 1.0
  ratio_clip: 0.2
  value_clip: 0.2
  clip_predicted_values: True
  entropy_loss_scale: 0.01
  value_loss_scale: 1.0
  kl_threshold: 0.0
  rewards_shaper_scale: 0.1
  time_limit_bootstrap: False
  # logging and checkpoint
  experiment:
    directory: "single_arm_reach_direct"
    experiment_name: ""
    write_interval: auto
    checkpoint_interval: auto


# Sequential trainer
# https://skrl.readthedocs.io/en/latest/api/trainers/sequential.html
trainer:
  class: SequentialTrainer
  timesteps: 36000
  environment_info: log
```

**Comment by KevinGeng3** [comment]
> Hi @KevinGeng3
> 
> In the [Isaac Lab examples of the **skrl** docs](https://skrl.readthedocs.io/en/latest/intro/examples.html#isaac-lab-environments) you can find the appropriate mapping between _rl_games_ and _skrl_ configurations.
> 
> This would be the mapping based on the _rl_games_ configuration you present. Please, adjust `mini_batches` according to the number of environments you are training on
> 
> ```yaml
> seed: 42
> 
> 
> # Models are instantiated using skrl's model instantiator utility
> # https://skrl.readthedocs.io/en/latest/api/utils/model_instantiators.html
> models:
>   separate: False
>   policy:  # see gaussian_model parameters
>     class: GaussianMixin
>     clip_actions: False
>     clip_log_std: True
>     min_log_std: -20.0
>     max_log_std: 2.0
>     initial_log_std: 0.0
>     network:
>       - name: net
>         input: STATES
>         layers: [64, 64]
>         activations: elu
>     output: ACTIONS
>   value:  # see deterministic_model parameters
>     class: DeterministicMixin
>     clip_actions: False
>     network:
>       - name: net
>         input: STATES
>         layers: [64, 64]
>         activations: elu
>     output: ONE
> 
> 
> # Rollout memory
> # https://skrl.readthedocs.io/en/latest/api/memories/random.html
> memory:
>   class: RandomMemory
>   memory_size: -1  # automatically determined (same as agent:rollouts)
> 
> 
> # PPO agent configuration (field names are from PPO_DEFAULT_CONFIG)
> # https://skrl.readthedocs.io/en/latest/api/agents/ppo.html
> agent:
>   class: PPO
>   rollouts: 36
>   learning_epochs: 5
>   mini_batches: 36 * NUM_ENVS / 73728  # horizon_length * num_actors / minibatch_size
>   discount_factor: 0.99
>   lambda: 0.95
>   learning_rate: 5.0e-04
>   learning_rate_scheduler: KLAdaptiveLR
>   learning_rate_scheduler_kwargs:
>     kl_threshold: 0.008
>   state_preprocessor: RunningStandardScaler
>   state_preprocessor_kwargs: null
>   value_preprocessor: RunningStandardScaler
>   value_preprocessor_kwargs: null
>   random_timesteps: 0
>   learning_starts: 0
>   grad_norm_clip: 1.0
>   ratio_clip: 0.2
>   value_clip: 0.2
>   clip_predicted_values: True
>   entropy_loss_scale: 0.01
>   value_loss_scale: 1.0
>   kl_threshold: 0.0
>   rewards_shaper_scale: 0.1
>   time_limit_bootstrap: False
>   # logging and checkpoint
>   experiment:
>     directory: "single_arm_reach_direct"
>     experiment_name: ""
>     write_interval: auto
>     checkpoint_interval: auto
> 
> 
> # Sequential trainer
> # https://skrl.readthedocs.io/en/latest/api/trainers/sequential.html
> trainer:
>   class: SequentialTrainer
>   timesteps: 36000
>   environment_info: log
> ```

Hi, thanks for your advice! But I think I have already adjusted mini_batches according to the number of environments(4096), so mini_batches = 2 is what I need. 
Are there any differences in the code implementation of PPO between RL-Games and SKRL? I roughly compared them and didn't find any obvious differences.

**Comment by KevinGeng3** [comment]
> It looks like setting time_limit_bootstrap to True can make training better for me

Thanks for your advice! I will have a try.

**Comment by Toni-SM** [comment]
> Hi, thanks for your advice! But I think I have already adjusted mini_batches according to the number of environments(4096), so mini_batches = 2 is what I need. Are there any differences in the code implementation of PPO between RL-Games and SKRL? I roughly compared them and didn't find any obvious differences.

Hi @KevinGeng3 

Please, note that not only `mini_batches` is different in the original mapping you provided (left) compared to the new one (right)
![image](https://github.com/user-attachments/assets/dd057c98-9214-4753-b02b-932582aa3437)

Regarding the differences in the code implementation of PPO between *rl_games* and *skrl*, there are 3 features in *rl_games* that are not included in *skrl* at the moment:
* bound loss regularization 
* central value training
* "legacy" computation of the adaptive KL learning rate scheduler

**Comment by KevinGeng3** [comment]
> > Hi, thanks for your advice! But I think I have already adjusted mini_batches according to the number of environments(4096), so mini_batches = 2 is what I need. Are there any differences in the code implementation of PPO between RL-Games and SKRL? I roughly compared them and didn't find any obvious differences.
> 
> Hi @KevinGeng3
> 
> Please, note that not only `mini_batches` is different in the original mapping you provided (left) compared to the new one (right) ![image](https://private-user-images.githubusercontent.com/22400377/397512303-dd057c98-9214-4753-b02b-932582aa3437.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUwMDI2ODgsIm5iZiI6MTczNTAwMjM4OCwicGF0aCI6Ii8yMjQwMDM3Ny8zOTc1MTIzMDMtZGQwNTdjOTgtOTIxNC00NzUzLWIwMmItOTMyNTgyYWEzNDM3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI0VDAxMDYyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUwZWY1M2M5ZmUyOWIwYmMyNWRmZTMyMzVmZDIwMzIwMWE2N2I0YTg2NmE0NzJhNjRkMGUwZGMxMjUyOGUzMTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.2j6XrJ67yVemmYBdZsR6pviS2rBCPLaRuCK_SwvBOOc)
> 
> Regarding the differences in the code implementation of PPO between _rl_games_ and _skrl_, there are 3 features in _rl_games_ that are not included in _skrl_ at the moment:
> 
> * bound loss regularization
> * central value training
> * "legacy" computation of the adaptive KL learning rate scheduler

Thank you!
